apis/v1alpha1/types.go

```diff
@@ -114,3 +114,98 @@ type ProviderConfigUsageList struct {
 	metav1.ListMeta `json:"metadata,omitempty"`
 	Items           []ProviderConfigUsage `json:"items"`
 }
+
+// +kubebuilder:object:root=true
+
+// A ObservedObjectCollection is a provider Kubernetes API type
+// +kubebuilder:subresource:status
+// +kubebuilder:printcolumn:name="KIND",type="string",JSONPath=".spec.kind"
+// +kubebuilder:printcolumn:name="APIVERSION",type="string",JSONPath=".spec.apiVersion",priority=1
+// +kubebuilder:printcolumn:name="PROVIDERCONFIG",type="string",JSONPath=".spec.providerConfigRef.name"
+// +kubebuilder:printcolumn:name="SYNCED",type="string",JSONPath=".status.conditions[?(@.type=='Synced')].status"
+// +kubebuilder:printcolumn:name="READY",type="string",JSONPath=".status.conditions[?(@.type=='Ready')].status"
+// +kubebuilder:printcolumn:name="AGE",type="date",JSONPath=".metadata.creationTimestamp"
+// +kubebuilder:resource:scope=Cluster,categories={crossplane,managed,kubernetes}
+type ObservedObjectCollection struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	Spec              ObservedObjectCollectionSpec   `json:"spec"`
+	Status            ObservedObjectCollectionStatus `json:"status,omitempty"`
+}
+
+// +kubebuilder:object:root=true
+
+// ObservedObjectCollectionList contains a list of ObservedObjectCollection
+type ObservedObjectCollectionList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []ObservedObjectCollection `json:"items"`
+}
+
+// ObservedObjectCollectionSpec defines the desired state of ObservedObjectCollection
+type ObservedObjectCollectionSpec struct {
+
+	// APIVersion of objects that should be matched by the selector
+	// +kubebuilder:validation:MinLength:=1
+	APIVersion string `json:"apiVersion"`
+
+	// Kind of objects that should be matched by the selector
+	// +kubebuilder:validation:MinLength:=1
+	Kind string `json:"kind"`
```

Comment:
I briefly tried to find if this was explored/concluded in the design discussion, e.g. https://github.com/crossplane-contrib/provider-kubernetes/issues/209#issuecomment-2016652897, but didn't find it after very minimal searching, so thank you for educating me @pedjak 😂 

Is the `Kind` field optional at all? i.e. does this limit the user to exactly one `Kind` of objects that this collection can find and observe? Is the expectation here that the user knows all the Kinds that an `Object` may end up creating, or that descendants of their `Object` may create (e.g., `Deployment` -> `ReplicaSet` -> `Pods`), and therefore they create an `ObservedObjectCollection` for each `Kind` they anticipate being created within that graph?

Reply:
I think the conclusion was to start with the simplest implementation which requires `Kind` and then see if we can add the option to do something more complicated later.  I think it would be really useful to be able to say "find all of the resources in the owner tree using resource X as the root" but I'm not sure how we would implement that efficiently.  The GC is able to do it and we may or may not want to/be able to replicate that functionality.  It may also be possible to write a Composition Function that has some knowledge of what a particular resource's child `Kind`s would be and use the extra-resources functionality to walk the tree and create `Object`s or `ObservedObjjectCollection`s as needed.

Reply:
I'd definitely prefer to keep it simple and require `kind`.

This is informed by a real use case from the team at Nike, which was CronJob -> Jobs and Deployment -> Pods. In both cases one input object (e.g. CronJob) will only produce one kind of object (i.e. Job) that the function is interested in.

As you mention @jbw976, if there's a reasonably small set of kinds you could always create one ObservedObjectCollection each.